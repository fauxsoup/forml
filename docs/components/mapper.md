## Members

<dl>
<dt><a href="#ArrayComponent">ArrayComponent</a></dt>
<dd><p>A wrapper for the array that utilizes the useArrayItems hook
to keep track of its children. Invokes the <a href="/docs/context">Context&#39;s</a></p>
<p>Usage:</p>
<pre><code class="language-jsx">const value = [&#39;a&#39;, &#39;b&#39;, &#39;c&#39;];
const form = {type: &#39;array&#39;, items: {type: &#39;string&#39;}};
const error = null;
&lt;ArrayComponent value={value} form={form} error={error} /&gt;</code></pre>
</dd>
</dl>

## Constants

<dl>
<dt><a href="#mapperTypes">mapperTypes</a></dt>
<dd><p>We&#39;re strictly a keyed collection of elements, so generate our
PropTypes.shape from an array of keys</p>
</dd>
</dl>

<a name="ArrayComponent"></a>

## ArrayComponent
A wrapper for the array that utilizes the useArrayItems hook
to keep track of its children. Invokes the [Context's](/docs/context)

Usage:

```jsx
const value = ['a', 'b', 'c'];
const form = {type: 'array', items: {type: 'string'}};
const error = null;
<ArrayComponent value={value} form={form} error={error} />
```

**Kind**: global variable  
**Component**: ArrayComponent  
<a name="mapperTypes"></a>

## mapperTypes
We're strictly a keyed collection of elements, so generate our
PropTypes.shape from an array of keys

**Kind**: global constant  
