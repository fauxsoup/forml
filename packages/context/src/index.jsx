import { createContext } from 'react';

/**
 * @typedef {Object} ModelMethods
 * @property {function} getValue - Retrieve a value from the model
 * @property {function} setValue - Update a value in the model
 * @property {function} getError - Retrieve a validation error from the model
 * @property {function} onChange - The callback to use when an event triggers
 * a change in the model.
 */

export const ModelContext = createContext({
    mapper: {},
    version: 0,
});
export const RenderingContext = createContext({
    mapper: {},
    decorator: {},
    localizer: {},
});

export default ModelContext;
