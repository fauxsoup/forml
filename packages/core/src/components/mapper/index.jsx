/**
 * @namespace forml.mapper
 */
import t from 'prop-types';

import Array from './array';
import Checkbox from './checkbox';
import Date from './date';
import DateTime from './datetime';
import makeDynamic from './dynamic';
import FieldSet from './fieldset';
import File from './file';
import Help from './help';
import Integer from './integer';
import Multiselect from './multiselect';
import Null from './null';
import Number from './number';
import Select from './select';
import Tabs from './tabs';
import Text from './text';
import TextArea from './textarea';

export function defaultMapper(SchemaForm = null) {
    return {
        array: Array,
        checkbox: Checkbox,
        date: Date,
        datetime: DateTime,
        dynamic: makeDynamic(SchemaForm),
        fieldset: FieldSet,
        help: Help,
        integer: Integer,
        multiselect: Multiselect,
        null: Null,
        number: Number,
        password: Text,
        select: Select,
        tabs: Tabs,
        text: Text,
        textarea: TextArea,
        tuple: FieldSet,
        file: File,
    };
}

/**
 * We're strictly a keyed collection of elements, so generate our
 * PropTypes.shape from an array of keys
 */
export const mapperTypes = [
    'array',
    'checkbox',
    'date',
    'datetime',
    'fieldset',
    'help',
    'integer',
    'multiselect',
    'null',
    'number',
    'password',
    'select',
    'tabs',
    'text',
    'textarea',
    'tuple',
];

export const mapperShape = t.shape({
    array: t.elementType,
    checkbox: t.elementType,
    date: t.elementType,
    datetime: t.elementType,
    fieldset: t.elementType,
    help: t.elementType,
    integer: t.elementType,
    multiselect: t.elementType,
    null: t.elementType,
    number: t.elementType,
    password: t.elementType,
    select: t.elementType,
    tabs: t.elementType,
    text: t.elementType,
    textarea: t.elementType,
    tuple: t.elementType,
});

export function getMapper(mapper = {}, SchemaForm = null) {
    return { ...defaultMapper(SchemaForm), ...mapper };
}
