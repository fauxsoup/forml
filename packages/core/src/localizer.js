/**
 * @typedef Localizer
 * @property {function():string} getLocalizedString - A string localizer
 * @property {function():string} getLocalizedDate - A date/time localizer
 * @property {function():string} getLocalizedNumber - A number localizer
 */

/**
 * The default localizer, which performs no operations.
 * @return {Localizer}
 */
export function defaultLocalizer() {
    function noop(id) {
        return id;
    }

    return {
        getLocalizedString: noop,
        getLocalizedNumber: noop,
        getLocalizedDate: noop,
    };
}

/**
 * Construct a localizer, using noop for any missing methods
 * @return {Localizer}
 */
export function getLocalizer(template) {
    return { ...defaultLocalizer(), ...template };
}
