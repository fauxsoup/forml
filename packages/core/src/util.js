import shortid from 'shortid';
import Ajv from 'ajv';
import objectPath from 'objectpath';
import {useMemo} from 'react';

/**
 * @namespace forml.util
 */

/**
 * @typedef {Function} ValueGetter
 * @param {Array<string|number>} keys - Object path to get
 * @return {*} - The value found at keys
 * @description Searches the enclosed model along the path of keys
 */

/**
 * @typedef {Function} ValueSetter
 * @param {Array<string|number>} keys - Object path to set
 * @param {*} value - Value to set at keys
 * @return {*} - The new model
 * @description Updates the enclosed model with value along the path of keys
 */

/**
 * Returns a default value for schema. If the schema defines a default value,
 * that will be returned. If no default is specified, an "empty" value of the
 * specified type will be returned.
 * @param {object} schema
 * @return {*}
 */
export function defaultForSchema(schema) {
  if (schema.default !== undefined) {
    return schema.default;
  }

  return buildSchema(schema);

  function buildSchema(schema) {
    const type = getPreferredType(schema.type);
    let base = undefined;
    switch (type) {
      case 'array':
        base = [];
        if (Array.isArray(schema.items)) {
          for (const item of schema.items) {
            base.push(defaultForSchema(item));
          }
        }
        break;
      case 'object':
        const required = schema.required || [];
        base = {};
        for (const property of required) {
          const item = defaultForSchema(schema.properties[property]);
          base[property] = item;
        }
        break;
      case 'string':
        base = '';
        break;
      case 'number':
        base = 0.0;
        break;
      case 'integer':
        base = 0;
        break;
      case 'boolean':
        base = false;
        break;
      case 'null':
        base = null;
        break;
      default:
        // throw new Error(`Unhandled defaultForSchema type: ${type}`);
        base = undefined;
    }

    return assertType(schema, base);
  }
}

export function randomForSchema(schema) {
  if (schema.default !== undefined) {
    return schema.default;
  }

  return buildSchema(schema);

  function buildSchema(schema) {
    const type = getPreferredType(schema.type);
    let base = undefined;

    switch (type) {
      case 'array':
        base = [];
        if (Array.isArray(schema.items)) {
          for (const item of schema.items) {
            base.push(randomForSchema(item));
          }
        } else {
          base.push(randomForSchema(schema.items));
        }
        break;
      case 'object':
        base = {};
        for (const property in schema.properties) {
          const item = randomForSchema(schema.properties[property]);
          base[property] = item;
        }
        break;
      case 'string':
        base = shortid();
        if (schema.format === 'date') {
          const date = new Date(
              Math.floor(Math.random() * new Date().getTime()),
          );
          base = date
              .toISOString()
              .match(/([0-9]{4}-[0-9]{2}-[0-9]{2})/)[1];
        } else if (schema.format === 'date-time') {
          base = new Date(
              Math.floor(Math.random() * new Date().getTime()),
          ).toISOString();
        }
        break;
      case 'number':
        base = Math.random();
        break;
      case 'integer':
        base = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
        break;
      case 'boolean':
        base = Math.random() > 0.5 ? true : false;
        break;
      case 'null':
        base = null;
        break;
      default:
        // throw new Error(`Unhandled randomForSchema type: ${type}`);
        base = undefined;
    }

    if (schema.enum) {
      const index = Math.floor(Math.random() * (schema.enum.length - 1));
      base = schema.enum[index];
    }

    return assertType(schema, base);
  }
}

/**
 * Return the first type from a type list
 * @arg {string[]} types
 */
export function getPreferredType(types) {
  const ignoredTypes = new Set(['null']);

  let index = 0;

  if (!Array.isArray(types)) return types;

  // Skip ignored types
  while (ignoredTypes.has(types[index])) index++;

  // If we've run past the end, just use the first type
  if (index >= types.length) index = 0;

  return types[index];
}

/**
 * @arg {*} model
 * @arg {object} schema
 * @return {ValueGetter}
 */
export function valueGetter(model, schema) {
  function get(keys) {
    if (!Array.isArray(keys)) {
      keys = [keys];
    }

    if (model === undefined) {
      model = defaultForSchema(schema);
    }

    if (keys.length === 0) return model;

    let current = model;
    let currentSchema = schema;

    for (let i = 0; i < keys.length; ++i) {
      const key = keys[i];
      currentSchema = getNextSchema(currentSchema, key);
      current = getNextValue(currentSchema, current, key);
    }

    return assertType(currentSchema, current);
  }

  return get;
}

export function assertType(schema, value) {
  const preferred = getPreferredType(schema.type);
  const allowed = new Set(
        Array.isArray(schema.type) ? schema.type : [schema.type],
  );
  const type = getTypeOf(schema, value);

  if (allowed.has('null') && !value) {
    return null;
  } else if (preferred === 'integer') {
    if (allowed.has(type)) {
      return value;
    } else {
      if (type === 'number' && Number.isInteger(value)) return value;
      else if (type === 'number') return Math.floor(value);
      else if (value === '') return value;
      else if (value === '-') return value;
      else if (type === 'string') return parseInt(value);
      else return defaultForSchema(schema);
    }
  } else if (preferred === 'number') {
    if (allowed.has(type)) {
      return value;
    } else if (type === 'string') {
      if (value === '') return value;
      else if (value === '-') return value;
      else if (/\.$/.test(value) && !/^[^.]+\.[^.]+\.$/.test(value)) {
        return value;
      } else return parseFloat(value);
    } else {
      return defaultForSchema(schema);
    }
  } else if (preferred === 'string' && type === 'number') {
    return value.toString();
  } else if (preferred != type && !value) {
    return defaultForSchema(schema);
  } else if (allowed.has(type)) {
    return value;
  } else {
    return defaultForSchema(schema);
  }
}

/**
 * @arg {object} errors - Map of object path to error
 * @return {Function}
 */
export function errorGetter(errors) {
  return function(keys) {
    const key = objectPath.stringify(keys);
    return errors[key];
  };
}

/**
 * Walk a key path along a schema tree and model, updating the model according
 * to the schema along the way, until reaching the final key, whereupon we set
 * the supplied value. Returns the updated model or value set.
 * @arg {Array<string|number>} keys - The object path to walk
 * @arg {*} model - The model to update
 * @arg {object} schema - The schema definition for the model
 * @arg {*} value - The final value to write
 * @arg {number} depth - The current depth of recursion
 * @return {*}
 */
function updateAndClone(keys, model, schema, value, depth = 0) {
  if (keys.length === 0) {
    return assertType(schema, value);
  }

  const [next, ...rest] = keys;
  const nextSchema = getNextSchema(schema, next);
  const nextModel = updateAndClone(
      rest,
      getNextValue(nextSchema, model, next),
      nextSchema,
      value,
      depth + 1,
  );

  if (getPreferredType(schema.type) === 'array') {
    const firstSlice = model.slice(0, next);
    const lastSlice = model.slice(next + 1);

    while (firstSlice.length < next) {
      firstSlice.push(
          defaultForSchema(getNextSchema(schema, firstSlice.length)),
      );
    }

    const result = [...firstSlice, nextModel, ...lastSlice];
    return result;
  }

  if (getPreferredType(schema.type) === 'object') {
    if (isRequired(schema, next) || isSaturated(nextModel)) {
      const result = {...model, [next]: nextModel};
      return result;
    } else {
      const {[next]: oldModel, ...nextResult} = model;
      return nextResult;
    }
  }

  throw new Error('Bad ObjectPath');
}

export function isRequired(schema, key) {
  if (schema.required) {
    return schema.required.includes(key);
  } else {
    return false;
  }
}

export function isSaturated(value) {
  if (Array.isArray(value)) {
    return value.length > 0;
  } else if (typeof value === 'object') {
    if (value === null) {
      return false;
    } else {
      return Object.keys(value).length > 0;
    }
  } else {
    return !!value;
  }
}

/**
 * Set a value using a hook setter
 * @arg {*} model
 * @arg {object} schema
 * @arg {Function} setModel
 */
export function valueSetter(model, schema) {
  function set(keys, value) {
    if (!Array.isArray(keys)) {
      keys = [keys];
    }

    if (model === undefined) {
      model = defaultForSchema(schema);
    }

    const newModel = updateAndClone(keys, model, schema, value);
    return assertType(schema, newModel);
  }

  return set;
}

/**
 * Walk the schema along the path of keys and return the last entry visited
 * @arg {Array<string|number>} keys
 * @arg {object} schema
 * @return {object}
 */
export function findSchema(keys, schema) {
  if (keys.length === 0) return schema;

  for (let i = 0; i < keys.length; ++i) {
    const key = keys[i];
    schema = getNextSchema(schema, key);
  }

  return schema;
}

/**
 * Return the child schema defined by key in this schema
 * @arg {object} schema
 * @arg {string|number} key
 */
export function getNextSchema(schema, key) {
  if (schema.type === 'array') {
    if (Array.isArray(schema.items)) {
      return schema.items[key];
    }
    return schema.items;
  }

  if (schema.type === 'object') {
    if (key in schema.properties) {
      return schema.properties[key];
    }

    if (schema.additionalProperties) {
      return schema.additionalProperties;
    }
  }
}

/**
 * Return the child model defined by key in this model, or if it is undefined
 * the default value for the schema.
 * @arg {object} schema
 * @arg {*} value
 * @arg {string|number} key
 */
export function getNextValue(schema, value, key) {
  if (!value) return defaultForSchema(schema);

  if (value[key] === undefined) {
    return defaultForSchema(schema);
  }

  return assertType(schema, value[key]);
}

export function getTypeOf(schema, value) {
  if (value === undefined) return getPreferredType(schema.type);
  else if (value === null) return 'null';
  else if (Array.isArray(value)) return 'array';
  else return typeof value;
}

/**
 * Invoke a function for every form in forms
 * @arg {object[]} forms - The forms tree to visit
 * @arg {Function} visitor - The visitor function to invoke
 */
export function traverseForm(forms, visit) {
  if (!Array.isArray(forms)) forms = [forms];

  for (const form of forms) {
    visit(form);

    if (form.items) traverseForm(form.items, visit);
  }
}

/**
 * A hook that uses a memoized Ajv instance and a compiled version of the schema
 * to produce a validate function.
 * @arg object schema - The schema to compile for validation
 */
const ajv = new Ajv({allErrors: true});
export function useValidator(schema) {
  const compiled = useMemo(() => {
    try {
      const validator = ajv.compile(schema);
      return validator;
    } catch (err) {
      const validator = () => false;
      validator.errors = [{dataPath: '.', message: 'invalid schema'}];
      return validator;
    }
  }, [schema]);

  function validate(model) {
    const valid = compiled(model);
    const {errors} = compiled;
    return {valid, errors};
  }

  return validate;
}

/**
 * A copy of clone that works on ES6 modules. Does not actually clone
 * primitives.
 * @arg * value - The value to clone
 */
export function clone(value) {
  switch (typeof value) {
    case 'string':
    case 'number':
    case 'boolean':
    case 'undefined': {
      return value;
    }

    case 'object': {
      if (Array.isArray(value)) {
        return value.map(clone);
      }

      const result = {};
      for (const key in value) {
        result[key] = clone(value[key]);
      }
      return result;
    }

    default: {
      return value;
    }
  }
}
