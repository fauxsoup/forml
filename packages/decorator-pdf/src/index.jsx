import Checkbox from './checkbox';
import FieldSet from './fieldset';
import Group from './group';
import Label from './label';
import Text from './text';
import * as Input from './input';
import * as Arrays from './arrays';
import * as Tabs from './tabs';

export { Arrays, Checkbox, FieldSet, Group, Input, Label, Tabs, Text };
