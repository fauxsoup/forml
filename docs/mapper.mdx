---
name: Mapper
route: /mapper
---

import {SchemaForm, decorators} from '../src';
import {Playground, Props} from 'docz';

# SchemaForm

This is an example:

<Props of={SchemaForm} />

<SchemaForm form={['*']} schema={{type: 'string'}} decorator={decorators.mui} model="Hello, friend" />
