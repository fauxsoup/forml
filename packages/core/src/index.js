/**
 * @namespace forml
 */

import * as util from "./util";
import * as constants from "./constants";

export * from "./schema-form";
export * from "./schema-field";

export * from "@forml/ware";

export { constants, util };
