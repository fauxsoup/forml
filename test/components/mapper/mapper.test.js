import { wrapInTestContext } from 'react-dnd-test-utils';
import { DndProvider } from 'react-dnd';
import Context from '../../../src/context';
import { getDecorator } from '../../../src/components/decorator';
import { getMapper } from '../../../src/components/mapper';
import { getLocalizer } from '../../../src/localizer';
import * as util from '../../../src/util';
import { createElement as h } from 'react';
import renderer from 'react-test-renderer';

const mapperTypes = [
    'array',
    'checkbox',
    'date',
    'datetime',
    'fieldset',
    'help',
    'integer',
    'multiselect',
    'null',
    'number',
    'password',
    'select',
    'tabs',
    'text',
    'textarea',
    'tuple',
];

describe('defaultMapper', function() {
    const mapper = getMapper({});
    mapperTypes.forEach((type) => {
        test(`includes entry for ${type}`, function() {
            expect(mapper[type]).not.toBeUndefined();
        })
    })
})

describe('getMapper', function() {
    const text = jest.fn((props) => 'text');
    const custom = jest.fn((propx) => 'custom');
    const mapper = getMapper({
        text,
        custom,
    });

    mapperTypes.forEach((type) => {
        test(`includes entry for ${type}`, function() {
            expect(mapper[type]).not.toBeUndefined();
        })
    });

    test('accepts custom field types', function() {
        expect(mapper.custom).not.toBeUndefined();
    });

    test('can override built-in types', function() {
        expect(mapper.text).toBe(text);
    })
})


describe('rendering', function() {
    const mapper = getMapper({});
    const localizer = getLocalizer({});
    const decorator = getDecorator({});
    const context = { mapper, localizer, decorator }

    mapperTypes.forEach(rendersElement('label', ['tabs', 'array', 'tuple', 'help', 'fieldset', 'null']))

    function rendersElement(element, excluded = []) {
        return function callback(type) {
            if (excluded.includes(type)) return;

            test(`includes a ${element} for ${type} if supplied`, function() {
                const schema = { type: 'null' };
                const form = { type, schema, key: [], title: 'test', titleMap: [], items: [], tabs: [] };
                const getValue = jest.fn(util.valueGetter(null, schema));
                const setValue = jest.fn(util.valueSetter(null, schema));
                const mappedComponent = wrapInTestContext(mapper[type]);
                const component = renderer.create(
                    h(
                        Context.Provider,
                        { value: { ...context, getValue, setValue } },
                        h(mappedComponent, { schema, form })
                    )
                ).toJSON();

                expect(component).toMatchSnapshot();
                expect(component.children.find((child) => child.type === element)).not.toBeUndefined();
            });
        }
    }
});
