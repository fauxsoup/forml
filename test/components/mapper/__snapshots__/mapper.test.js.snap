// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering includes a label for checkbox if supplied 1`] = `
<div>
  <label>
    <input
      onChange={[Function]}
      type="checkbox"
    />
    test
  </label>
</div>
`;

exports[`rendering includes a label for date if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "date",
      }
    }
  >
    test
  </label>
  <input
    onChange={[Function]}
    type="date"
  />
</div>
`;

exports[`rendering includes a label for datetime if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "datetime",
      }
    }
  >
    test
  </label>
  <input
    onChange={[Function]}
    type="datetime"
  />
</div>
`;

exports[`rendering includes a label for integer if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "integer",
      }
    }
  >
    test
  </label>
  <input
    onChange={[Function]}
    type="integer"
  />
</div>
`;

exports[`rendering includes a label for multiselect if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "multiselect",
      }
    }
  >
    test
  </label>
  <select
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "multiselect",
      }
    }
    multiple={true}
    onChange={[Function]}
  />
</div>
`;

exports[`rendering includes a label for number if supplied 1`] = `
<div>
  <label
    focused={false}
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "number",
      }
    }
    htmlFor=""
  >
    test
  </label>
  <input
    onChange={[Function]}
    type="number"
  />
</div>
`;

exports[`rendering includes a label for password if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "password",
      }
    }
  >
    test
  </label>
  <input
    onChange={[Function]}
    type="password"
  />
</div>
`;

exports[`rendering includes a label for select if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "select",
      }
    }
  >
    test
  </label>
  <select
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "select",
      }
    }
    onChange={[Function]}
  />
</div>
`;

exports[`rendering includes a label for text if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "text",
      }
    }
  >
    test
  </label>
  <input
    onChange={[Function]}
    type="text"
  />
</div>
`;

exports[`rendering includes a label for textarea if supplied 1`] = `
<div>
  <label
    form={
      Object {
        "items": Array [],
        "key": Array [],
        "schema": Object {
          "type": "null",
        },
        "tabs": Array [],
        "title": "test",
        "titleMap": Array [],
        "type": "textarea",
      }
    }
  >
    test
  </label>
  <input
    onChange={[Function]}
    type="textarea"
  />
</div>
`;
